infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/FsmTaskManager.java

```diff
@@ -0,0 +1,275 @@
+        }
+    }
+
+    /**
+     * Add event to in-memory session->event maps. This is to ensure the order of event processing for a given session.
+     * Currently the "delay" is non-zero for only the replicating FSM.
+     */
+    private <E> void addEventToSessionEventMap(E event, FsmEventType fsm, long delay) {
+        if (fsm.equals(FsmEventType.LogReplicationRuntimeEvent)) {
+            LogReplicationSession session = ((LogReplicationRuntimeEvent) event).getRuntimeFsm().getSession();
+            sessionToRuntimeEventIdMap.putIfAbsent(session, new LinkedList<>());
+            // makes the value part of the map thread safe.
+            sessionToRuntimeEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                eventList.add(((LogReplicationRuntimeEvent) event).getEventId());
+                return eventList;
+            });
+        } else if (fsm.equals(FsmEventType.LogReplicationEvent)){
+            LogReplicationSession session = ((LogReplicationEvent) event).getReplicationFsm().getSession();
+            if (delay > 0) {
+                sessionToDelayedReplicationEventIdMap.putIfAbsent(session, new LinkedList<>());
+                // makes the value part of the map thread safe.
+                sessionToDelayedReplicationEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                    eventList.add(((LogReplicationEvent) event).getEventId());
+                    return eventList;
+                });
+            } else {
+                sessionToReplicationEventIdMap.putIfAbsent(session, new LinkedList<>());
+                // makes the value part of the map thread safe.
+                sessionToReplicationEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                    eventList.add(((LogReplicationEvent) event).getEventId());
+                    return eventList;
+                });
+            }
+        } else if (fsm.equals(FsmEventType.LogReplicationSinkEvent)){
+            LogReplicationSession session = ((LogReplicationSinkEvent) event).getSourceLeadershipManager().getSession();
+            sessionToSinkEventIdMap.putIfAbsent(session, new LinkedList<>());
+            // makes the value part of the map thread safe.
+            sessionToSinkEventIdMap.computeIfPresent(session, (sessionKey, eventList) -> {
+                eventList.add(((LogReplicationSinkEvent) event).getEventId());
+                return eventList;
+            });
+        } else {
+            log.warn("Ignored {}. Unexpected FSM event type", fsm);
+        }
+    }
+
+    private void processSinkTask(LogReplicationSinkEvent event) {
+        LogReplicationSession session = event.getSourceLeadershipManager().getSession();
+        synchronized (event.getSourceLeadershipManager()) {
+            while (!sessionToSinkEventIdMap.get(session).get(0).equals(event.getEventId())) {
+                try {
+                    event.getSourceLeadershipManager().wait();
```

Comment:
As there are 2 threads, this event is at max at the 2nd position.  Should we check if the 1st event belongs to the same session and if not, process this event?

Because otherwise, events across sessions will always get executed sequentially even though we have multiple threads, right?

We can discuss a bit more before making the changes.

Reply:
No, events across the sessions are parallelized. This wait happens only when the event is in the 2nd position for a particular session.
 If there are more than 2 threads, they'll go to the 3rd event, and say if it is from a different session and is in position 1 for that session, it gets processed.

we can certainly discuss this more..